File: src/test/java/frc/team6223/arsenalFramework/RateTest.java
package frc.team6223.arsenalFramework;

import frc.team6223.arsenalFramework.software.units.*;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

import static org.assertj.core.api.Assertions.*;

public class RateTest {

    private Rate<DistanceUnits, TimeUnits> rate;

    @BeforeEach
    public void init() {
        rate = new Velocity(new Distance(1.0, DistanceUnits.METERS), new Time(1.0, TimeUnits.SECONDS));
    }

    @Test
    @DisplayName("rate converts from m/s to m/ms correctly")
    public void downscaleRate() {
        assertThat(rate.rescaleBottomScalar(TimeUnits.MILLISECONDS)).isEqualTo(0.001);
    }

    @Test
    @DisplayName("rate converts from m/s to m/min correctly")
    public void upscaleRate() {
        assertThat(rate.rescaleBottomScalar(TimeUnits.MINUTE)).isEqualTo(60.0);
    }

    @Test
    @DisplayName("rate converts from non-unit rate to unit rate upon creation")
    public void unitRate() {
        assertThat(
          new Velocity(
            new Distance(1.0, DistanceUnits.METERS),
            new Time(5.0, TimeUnits.SECONDS)
          ).numericValue()).isEqualTo(0.20);
    }

    @Test
    @DisplayName("rate formats correctly via toString()")
    public void toStringTest() {
        assertThat(rate.toString()).isEqualTo("1.0 m/s");
    }

}

File: src/test/kotlin/frc/team6223/arsenalFramework/RateTest.kt
package frc.team6223.arsenalFramework

import frc.team6223.arsenalFramework.software.units.*
import io.kotlintest.matchers.plusOrMinus
import io.kotlintest.matchers.shouldBe
import io.kotlintest.specs.StringSpec

class RateTest: StringSpec() {
    init {
        val rate = Rate(Distance(1.0, DistanceUnits.METERS), Time(1.0, TimeUnits.SECONDS))
        "rate converts from m/s to m/ms correctly" {
            rate.rescaleBottomScalar(TimeUnits.MILLISECONDS) shouldBe 0.001

            rate.rescaleBottomRate(TimeUnits.MILLISECONDS).numericValue() shouldBe 0.001
        }

        "rate converts from m/s to m/min correctly" {
            rate.rescaleBottomScalar(TimeUnits.MINUTE) shouldBe (60.0 plusOrMinus .1)

            rate.rescaleBottomRate(TimeUnits.MINUTE).numericValue() shouldBe (60.0 plusOrMinus .1)
        }

        "rate converts from non-unit rate to a unit rate upon creation" {
            val rate = Rate(Distance(1.0, DistanceUnits.METERS), Time(5.0, TimeUnits.SECONDS))
            rate.numericValue() shouldBe 0.20
        }

        "rate formats correctly via toString()" {
            rate.toString() shouldBe "1.0 m/s"
        }
    }
}